backend/src/main/java/org/eclipse/tractusx/semantics/registry/service/GranularShellAccessHandler.java

```diff
@@ -181,16 +184,36 @@ private ShellVisibilityContext toShellVisibilityContext( Shell shell ) {
 
    private Set<ShellIdentifier> filterSpecificAssetIdsByTenantId( Set<ShellIdentifier> shellIdentifiers, ShellVisibilityCriteria visibilityCriteria ) {
       //noinspection SimplifyStreamApiCallChains
-      return shellIdentifiers.stream()
+      Set<ShellIdentifier> filterShellIdentifiers = new HashSet<>();
+
+      // filter and group by key
+      Map<String, List<ShellIdentifier>> filteredIdentifiersByKey = shellIdentifiers.stream()
             .filter( identifier -> identifier.getKey().equals( ShellIdentifier.GLOBAL_ASSET_ID_KEY )
-                                   || visibilityCriteria.visibleSpecificAssetIdNames().contains( identifier.getKey() ) )
+                  || visibilityCriteria.visibleSpecificAssetIdNames().containsKey( identifier.getKey() ) )
             //TODO: Do we need to clear the list of external subject Ids?
             .map( identifier -> {
                Optional.ofNullable( identifier.getExternalSubjectId() )
                      .ifPresent( extSubId -> extSubId.setKeys( Collections.emptySet() ) );
                return identifier;
             } )
-            .collect( Collectors.toSet() );
+            .collect( Collectors.groupingBy( ShellIdentifier::getKey ) );
+
+      // filter by value:
+      // In case of multiple specificAssetIds with same keys, the filtering rule is:
+      //1. if the value match with the given value from visibilityCriteria, then only this shell will be included and the others filtered out
+      //2. If the value not match with the given value from visibilityCriteria, all shells will be included because of the any visibility.
```

Comment:
Hi,
Shouldn't this be the other way around? I mean if we have only one value with the given `specificAssetId` name, then we can show it even when the `mandatorySpecificAssetId` collection does not include it, but as soon as the name is not unique, I think we are supposed to assume, that it is because of multiple entities using the same shell and we should avoid showing it unless the value is an exact match with the `mandatorySpecificAssetId` value.

Reply:
Done